Sources/SourceControl/RepositoryManager.swift

```diff
@@ -145,53 +157,84 @@ public class RepositoryManager {
     ///
     /// This will initiate a clone of the repository automatically, if necessary.
     public func lookup(repository: RepositorySpecifier, completion: @escaping LookupCompletion) {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
-            do {
-                let repo = try handle.open()
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+        concurrentQueue.async { 
+            // First look for the handle.
+            let handle = self.getHandle(repository: repository)
+            // Dispatch the action we want to take on the serial queue of the handle.
+            handle.serialQueue.async {
+                let result: LookupResult
+                switch handle.status {
+                case .available:
+                    // FIXME: Need to only do this for first lookup.
+                    result = try! LookupResult {
+                        // Fetch and update the repository when it is being looked up.
+                        // We need some mechanism to maybe not do this for pinned repos.
```

Comment:
FYI, this mechanism wouldn't be here: it would be on the resolver to not ask for pinned repos, I think.